diff --git a/disqus.js b/disqus.js
index 74d971d..deab80d 100644
--- a/disqus.js
+++ b/disqus.js
@@ -46,6 +46,9 @@ Drupal.behaviors.disqus = {
           if (settings.disqus.sso || false) {
             this.sso = settings.disqus.sso;
           }
+          if (typeof disqus_track_comments == 'function') {
+              this.callbacks.onNewComment = [function(params) { disqus_track_comments(params); }];
+          }
         };
 
         // Make the AJAX call to get the Disqus comments.
