diff --git a/disqus.js b/disqus.js
index 74d971d..6986aa0 100644
--- a/disqus.js
+++ b/disqus.js
@@ -46,6 +46,9 @@ Drupal.behaviors.disqus = {
           if (settings.disqus.sso || false) {
             this.sso = settings.disqus.sso;
           }
+          if (typeof Drupal.behaviors.comment_sources_disqus_track_comments == 'function') {
+            this.callbacks.onNewComment = [function(params) { Drupal.behaviors.comment_sources_disqus_track_comments(params); }];
+          }
         };
 
         // Make the AJAX call to get the Disqus comments.
diff --git a/disqus.module b/disqus.module
index 93fb84a..2bbfa06 100644
--- a/disqus.module
+++ b/disqus.module
@@ -192,7 +192,7 @@ function disqus_node_load($nodes, $types) {
   if (!empty($domain)) {
     // Load Disqus into the nodes.
     foreach ($nodes as &$node) {
-      if (!empty($disqustypes[$node->type])) {
+      if (!empty($disqustypes[$node->type]) || (module_exists('comment_sources') && comment_sources_has_comments($node,'disqus'))) {
         // Save the data to the node object.
         $node->disqus = array('domain' => $domain);
 
@@ -227,7 +227,23 @@ function disqus_node_load($nodes, $types) {
  * Implements hook_node_view().
  */
 function disqus_node_view($node, $view_mode) {
-  if (isset($node->disqus) && user_access('view disqus comments') && $node->disqus['status'] == 1) {
+  // Assume no disqus comments will be displayed.
+  $display_disqus = FALSE;
+  // If comment_sources module is not enabled, simply verify access and status.
+  if (!module_exists('comment_sources')) {
+    if (isset($node->disqus) && user_access('view disqus comments') && $node->disqus['status'] == 1) {
+      $display_disqus = TRUE;
+    }
+  }
+  // Otherwise, only display disqus comments if comment_sources says to.
+  else {
+    if (function_exists('comment_sources_should_display_comments')) {
+      if (comment_sources_should_display_comments($node, 'disqus')) {
+        $display_disqus = TRUE;
+      }
+    }
+  }
+  if ($display_disqus) {
     switch ($view_mode) {
       case 'full':
       	// Inject Disqus into the node object.
