Index: sites/all/modules/disqus/disqus.module
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- sites/all/modules/disqus/disqus.module	(revision )
+++ sites/all/modules/disqus/disqus.module	(revision )
@@ -227,7 +227,8 @@
  * Implements hook_node_view().
  */
 function disqus_node_view($node, $view_mode) {
-  if (isset($node->disqus) && user_access('view disqus comments') && $node->disqus['status'] == 1) {
+  if (isset($node->disqus) && user_access('view disqus comments') && $node->disqus['status'] == 1
+      && (($node->comment_count == 0))) {
     switch ($view_mode) {
       case 'full':
       	// Inject Disqus into the node object.
